riir-macro/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+                #[cfg_attr(not(feature = "span"), allow(unused_mut))]
+                let mut string = ident.to_string();
+                #[cfg(feature = "span")]
+                {
+                    fill_space(ident.span(), &mut string);
+                }
+                #[cfg(not(feature = "span"))]
+                {
+                    string.push(' ');
+                }
+                output.push_str(&string);
+            }
+            TokenTree::Punct(punct) => {
+                let mut string = punct.as_char().to_string();
+                #[cfg(feature = "span")]
+                {
+                    fill_space(punct.span(), &mut string);
+                }
+                #[cfg(not(feature = "span"))]
+                {
+                    use proc_macro2::Spacing;
+
+                    if punct.as_char() == '#' {
+                        if !was_hash {
+                            string = format!("\n{}", string);
+                        }
+                        was_hash = true;
+                    }
+                    if punct.spacing() == Spacing::Joint {
+                        string.push(' ');
+                    }
+                }
+                output.push_str(&string);
+            }
+            TokenTree::Literal(lit) => {
+                #[cfg_attr(not(feature = "span"), allow(unused_mut))]
+                let mut string = lit.to_string();
+                #[cfg(feature = "span")]
+                {
+                    fill_space(lit.span(), &mut string);
+                }
+                output.push_str(&string);
+            }
+            TokenTree::Group(group) => {
+                let mut string = String::new();
+                let (l, r) = match group.delimiter() {
+                    Delimiter::Parenthesis => ("(", ")"),
+                    Delimiter::Bracket => ("[", "]"),
+                    Delimiter::Brace => ("{", "}"),
+                    Delimiter::None => ("", ""),
+                };
+                string.push_str(l);
+                string.push_str(&print_ts(group.stream()));
+                string.push_str(r);
+                #[cfg(feature = "span")]
+                {
+                    fill_space(group.span(), &mut string);
+                }
+                output.push_str(&string);
+            }
+        }
+    }
+    output
+}
+
+#[cfg(feature = "span")]
+fn fill_space(span: Span, string: &mut String) {
+    let start_col = if span.end().line > span.start().line {
+        let lines = span.end().line - span.start().line;
+        string.extend((0..lines).map(|_| '\n'));
+        -1
+    } else {
+        (span.start().column + string.len()) as isize
+    };
+    let space_size = span.end().column as isize - start_col;
+    string.extend((0..space_size).map(|_| ' '));
+}
+
+#[cfg(test)]
+#[test]
+fn test() {
+    use std::{fs, path::PathBuf};
+
+    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test.txt");
```

Comment:
The test file was made external instead of using `quote!()`, because `quote!()` does not allow literal expressions with a `#`.